Jetpack Section: Threat Details File Context

momo-ozawa opened the pull request:
Part of #15190

### Description:
- Added an attributed string for file context in Threat Details

### Merge instructions:
- ⚠️ Merge after https://github.com/wordpress-mobile/WordPress-iOS/pull/15740

### To test:
1. Launch the app, tap on the My Site, tap on a site with Jetpack Scan and also has Threats
2. Tap on a threat cell for a threat that includes a file context
3. ✅ The file context can scroll horizontally
4. ✅ The file context styling should match the design specs  

![Screen Shot 2021-02-01 at 15 48 46](https://user-images.githubusercontent.com/6711616/106424166-0fa84700-64a5-11eb-8299-3be04e13f4b7.png)

### PR submission checklist:
- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Commit:
Display attributed file context string

Commit:
Embed threat context label in scrollview

Commit:
Update styling for threat details

Commit:
Fix typo

emilylaguna commented on WordPress/Classes/ViewRelated/Jetpack/Jetpack Scan/View Models/JetpackScanThreatViewModel.swift:
```diff
@@ -309,8 +311,128 @@ struct JetpackScanThreatViewModel {
 
+extension JetpackScanThreatViewModel {
+
+    private static func attributedString(for context: JetpackThreatContext?) -> NSAttributedString? {
```
How do you feel about extracting this into an extension on `JetpackThreatContext` that also takes a struct called something like `JetpackThreatContextRendererConfig`? Then the config struct can return the dictionaries for the NSAttributedString attributes?

momo-ozawa replied:
Extracted to a private `JetpackThreatContext` extension! ✨  de51d96

I decided to define a private extension inside `JetpackScanThreatViewModel` instead of WPKit, because attributed strings feel closer to a viewmodel than a data model.

emilylaguna commented on WordPress/Classes/ViewRelated/Jetpack/Jetpack Scan/View Models/JetpackScanThreatViewModel.swift:
```diff
@@ -309,8 +311,128 @@ struct JetpackScanThreatViewModel {
+        let numberFont = UIFont.monospacedSystemFont(ofSize: 13, weight: .bold)
+        let attrString = NSMutableAttributedString()
+
+        let biggestLen = String(context.lines.last?.lineNumber ?? 0).count
```
This name is pretty confusing. We could probably rename it to `lastNumberLength` which is slightly better, then add some documentation for it:

```        
        // Calculates the "longest" number string length
        // this will be used to pad the number column to make sure it's all the same size regardless of
        // the line number length
        //
        // ie: last line number is 1000, which is 4 chars
        // when processing line 50 we'll append 2 spaces to the end so it ends up being equal to 4
```

emilylaguna commented on WordPress/Classes/ViewRelated/Jetpack/Jetpack Scan/View Models/JetpackScanThreatViewModel.swift:
```diff
@@ -309,8 +311,128 @@ struct JetpackScanThreatViewModel {
+
+        let biggestLen = String(context.lines.last?.lineNumber ?? 0).count
+
+        let longestLine = context.lines.sorted { $0.contents.count > $1.contents.count}.first!
```
This line causes a crash if the context.lines is an empty array. We could probably guard or default instead of using a forced unwrap.

emilylaguna commented on WordPress/Classes/ViewRelated/Jetpack/Jetpack Scan/View Models/JetpackScanThreatViewModel.swift:
```diff
@@ -309,8 +311,128 @@ struct JetpackScanThreatViewModel {
+        let longestLine = context.lines.sorted { $0.contents.count > $1.contents.count}.first!
+        let longestLen = longestLine.contents.count
+
+        let paragraph = NSMutableParagraphStyle()
```
I think we can remove this since we don't actually change the paragraph style in the attributes.

```suggestion
```

emilylaguna commented on WordPress/Classes/ViewRelated/Jetpack/Jetpack Scan/View Models/JetpackScanThreatViewModel.swift:
```diff
@@ -309,8 +311,128 @@ struct JetpackScanThreatViewModel {
+            // Number attributed string
+
+            var numberStr = String(line.lineNumber)
+            let numberPadding = "".padding(toLength: biggestLen - numberStr.count,
```
```suggestion
            let numberPadding = String().padding(toLength: biggestLen - numberStr.count,
```

emilylaguna commented on WordPress/Classes/ViewRelated/Jetpack/Jetpack Scan/View Models/JetpackScanThreatViewModel.swift:
```diff
@@ -309,8 +311,128 @@ struct JetpackScanThreatViewModel {
+            ], range: NSRange(location: 0, length: numberStr.count))
+
+            attrString.append(numberAttr)
+            attrString.append(NSAttributedString(string: ""))
```
I don't think this is needed anymore. 

```suggestion
```

emilylaguna commented on WordPress/Classes/ViewRelated/Jetpack/Jetpack Scan/View Models/JetpackScanThreatViewModel.swift:
```diff
@@ -309,8 +311,128 @@ struct JetpackScanThreatViewModel {
+            // Contents attributed string
+
+            let contents = line.contents
+            let contentsPadding = "".padding(toLength: longestLen - contents.count,
```
```suggestion
            let contentsPadding = String().padding(toLength: longestLen - contents.count,
```

emilylaguna commented on WordPress/Classes/ViewRelated/Jetpack/Jetpack Scan/View Models/JetpackScanThreatViewModel.swift:
```diff
@@ -309,8 +311,128 @@ struct JetpackScanThreatViewModel {
+            let contentsPadding = "".padding(toLength: longestLen - contents.count,
+                                             withPad: Constants.noBreakSpace, startingAt: 0)
+
+            let padding = " "
```
I forgot what this was for, we can probably rename it to something like this or .. ?

We should also move this to Constants

```suggestion
            let columnSpacer = " "
```

emilylaguna commented on WordPress/Classes/ViewRelated/Jetpack/Jetpack Scan/View Models/JetpackScanThreatViewModel.swift:
```diff
@@ -309,8 +311,128 @@ struct JetpackScanThreatViewModel {
+        static let noBreakSpace = "\u{00a0}"
+
+        struct normal {
+            static let textColor = UIColor(rgb: 0x2C3337)
+            static let backgroundColor = UIColor(rgb: 0xF6F7F7)
+            static let numberTextColor = UIColor(rgb: 0x646970)
+            static let numberBackgroundColor = UIColor(rgb: 0xC3C4C7)
+        }
+
+        struct highlighted {
+            static let textColor = UIColor.white
+            static let backgroundColor = UIColor(rgb: 0xD63638)
+            static let numberBackgroundColor = UIColor(rgb: 0xF1DADA)
```
I think we can just convert these to normal UIColors so we don't need to add the hex to UIColor extension below. 

I normally use https://www.uicolor.io/ to help with this.

emilylaguna reviewed (changes requested):
Heya!! Nice work, I hit a few snags and left a few comments below.

Let me know your thoughts!

Commit:
Merge remote-tracking branch 'origin/develop' into task/15190-threat-context

Commit:
Clean up attributed string logic

Commit:
Convert hex colors to rgba

Commit:
Extract attributedString to a JetpackThreatContext extension

momo-ozawa commented:
@emilylaguna Thanks for reviewing! 
I've addressed your suggestions -- ready for another look! 😄

emilylaguna commented on WordPress/Classes/ViewRelated/Jetpack/Jetpack Scan/View Models/JetpackScanThreatViewModel.swift:
```diff
@@ -49,7 +50,37 @@ struct JetpackScanThreatViewModel {
+            ]
+        )
+
+        attributedFileContext = threat.context?.attributedString(with: contextConfig) ?? nil
```
Since `attributedString(with:`) already returns nil I don't think we need to default here. 

```suggestion
        attributedFileContext = threat.context?.attributedString(with: contextConfig)
```

emilylaguna reviewed (changes requested):
Heya! Looking awesome! I left 1 comment below, and I had 2 suggestions:

1. Not exactly related to this so we can touch on it later but can we change the file name font to the monospaced one? 
2. When loading my very large threats list on my site I am getting some stutters due to the attributed string. Can we defer the loading of this until we need it on the Threat Details view?

Commit:
Update file name font to monospaced

Commit:
Change font from bold to semibold

Commit:
Delete unnecessary default value

Commit:
Update attributed string to a lazy var

momo-ozawa commented:
@emilylaguna Back to you ✨ 

> Not exactly related to this so we can touch on it later but can we change the file name font to the monospaced one?

Fixed! https://github.com/wordpress-mobile/WordPress-iOS/pull/15749/commits/3da4bda7a3d5216871b6eb8ba55b424532a7c6b3

> Since attributedString(with:) already returns nil I don't think we need to default here.

Fixed! c00bc17

> When loading my very large threats list on my site I am getting some stutters due to the attributed string. Can we defer the loading of this until we need it on the Threat Details view?

Fixed! 06069c9

emilylaguna reviewed (approved):
LGTM 🚢
